src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -222,13 +223,34 @@ case object TChar extends Type {
     .map(s => s.substring(0, 1))
 }
 
-abstract class TNumeric extends Type
+object TNumeric {
+  def parseResultType(types: Set[TNumeric]): Type = {
+    if (types(TDouble) || (types(TFloat) && (types(TInt) || types(TLong))))
+      TDouble
+    else if (types(TInt) && types(TLong))
+      TLong
```

Comment:
`types` is a set.  If there's only one numeric type, that's what's returned.  If there are multiple longs but no ints, types is `Set(TLong)`, which will go into the third condition (and pass the assert).  The reason it's complex is that if you get types = `Set(TFloat, TLong)`, then you want to promote to Double